Include/Rover/TrialView.hpp

```diff
@@ -0,0 +1,270 @@
+      using Sample = S;
+
+      //! Random access constant iterator.
+      class ConstIterator {
+        public:
+
+          //! Creates a constant iterator.
+          /*
+            \param t Trial to read.
+            \param offset Offset from t.begin().
+          */
+          template<typename Trial>
+          ConstIterator(const Trial& t, std::ptrdiff_t offset);
+
+          //! Increments the iterator.
+          ConstIterator& operator ++();
+
+          //! Increments the iterator.
+          ConstIterator operator ++(int);
+
+          //! Decrements the iterator.
+          ConstIterator& operator --();
+
+          //! Decrements the iterator.
+          ConstIterator operator --(int);
+
+          //! Advances the iterator.
+          ConstIterator& operator +=(std::ptrdiff_t offset);
+
+          //! Moves the iterator back.
+          ConstIterator& operator -=(std::ptrdiff_t offset);
+
+          //! Returns an iterator that is the result of advancing this.
+          ConstIterator operator +(std::ptrdiff_t offset);
+
+          //! Returns an iterator that is the result of moving this back.
+          ConstIterator operator -(std::ptrdiff_t offset);
+
+          //! Returns the offset between two iterators.
+          std::ptrdiff_t operator -(ConstIterator other) const;
+
+          //! Checks whether two iterators point to the same Sample.
+          bool operator ==(ConstIterator other) const;
+
+          //! Checks whether two iterators do not point to the same Sample.
+          bool operator !=(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is to
+          //! the left of that pointed to by the second iterator.
+          bool operator <(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is not
+          //! to the right of that pointed to by the second iterator.
+          bool operator <=(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is to
+          //! the right of that pointed to by the second iterator.
+          bool operator >(ConstIterator other) const;
+
+          //! Checks whether the Sample pointed to by the first iterator is not
+          //! to the left of that pointed to by the second iterator.
+          bool operator >=(ConstIterator other) const;
+
+          //! Returns a reference to the Sample the iterator points to.
+          const Sample& operator *() const;
+
+          //! Returns a pointer to the Sample the iterator points to.
+          const Sample* operator ->() const;
+
+          //! Returns a reference to the Sample the iterator would point to
+          //! if it was advanced by the offset.
+          const Sample& operator [](std::ptrdiff_t offset) const;
+
+        private:
+          std::ptrdiff_t m_offset;
+          std::function<const Sample* (std::ptrdiff_t)> m_get_ptr;
```

Comment:
This implementation is too heavy, ```std::function``` is not a lightweight template by any means. Can you implement this instead by having a pointer to the ```TrialView``` and invoking its index operator? Something like:

```
  private:
    std::size_t m_index;
    const TrialView* m_view;

  ...

  const Sample* ConstIterator<...>::operator *() const {
    return (*m_view)[m_index];
  }
```

If you want to use ```std::function``` for type erasure on the ```Trial``` itself that's fine.

Reply:
m_get_ptr is used to erase Trial's type. Alternatively, I could erase its Type in TrialView itself, but I would need an std::function (or something equally complex) to implement operator[] then. I decided it was best to erase the type in ConstIterator instead as it involved less indirection.

Reply:
I would prefer to have one ```std::function``` per ```TrialView``` shared by all N ```ConstIterators```. So feel free to use ```std::function``` but only construct one instance of it per ```TrialView```.